File: 2024/01/1.clj
(ns aoc.2024.01.1
  (:require
   [aoc.common :refer [lines nums parse-input zip]]))

(let [[list1 list2] (->> (parse-input (lines nums))
                         (apply zip)
                         (map sort))
      diff (->> [list1 list2]
                (apply zip)
                (map #(abs (apply - %)))
                (apply +))
      freqs (frequencies list2)
      sim (->> list1
               (map #(* % (get freqs % 0)))
               (apply +))]

  (println "Part 1:" diff)
  (println "Part 2:" sim))

File: 2024/01/1.ts
import { $, difference, frequencies, ints, lines, map, pipe, readInput, sortNumeric, split, sum, zip } from '../../common'

const [list1, list2] = $(readInput(), lines, map(split(/\s+/)), zip, map(pipe(ints, sortNumeric())))
const diff = $([list1, list2], zip, map(difference), sum)
const freqs = frequencies(list2)
const sim = $(
  list1,
  map(v => v * freqs.get(v) || 0),
  sum
)

console.log('Part 1:', diff)
console.log('Part 2:', sim)
